### ISSUE TITLE
Stop creating unnecessary exceptions with MultiException

### ISSUE THREAD

**Issue by LukeButters** [eclipse/jetty.project#2580]
`MultiException` creates an Exception even if no exception is eventually thrown. This is not a great way to program in Java, as outlined in **Item 57 use exceptions only for exceptional conditions** in **Effective java**.

The problem for jetty is creating exceptions in Java is not free, rather they are computationally expensive. This means in any place that `MultiException` is used the code is slower than it should be. 

I think it is worth brining this up now before too much code is written to use `MultiException`.

The solution could be to change to using a `Builder` to record the exceptions and only create exception of we have at least one exception to throw.

I am happy to have a go at changing how `MultiException` works as long as a solution can be agreed upon.

**Comment by LukeButtersFunnelback** [comment]
We need to keep in mind that if `addSuppressed` is used we still need to avoid creating any extra exceptions when no exceptions would be thrown (or do something where we delay filling in the stack trace). I am looking at it from how we would use `addSuppressed` in https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-util/src/main/java/org/eclipse/jetty/util/MultiPartInputStreamParser.java#L465 which doesn't really lean itself to having a main exception and some other suppressed exceptions.

Aside we could change `MultiException` to set `writableStackTrace` in the parent constructor to `false`, and call `fillInStackTrace()` only when we want to throw the `MultiException`.

**Comment by LukeButtersFunnelback** [comment]
This diff shows a change which probably doesn't affect other files
```
diff --git a/jetty-util/src/main/java/org/eclipse/jetty/util/MultiException.java b/jetty-util/src/main/java/org/eclipse/jetty/util/MultiException.java
index c51b0e66ac..f45b4cdaa7 100644
--- a/jetty-util/src/main/java/org/eclipse/jetty/util/MultiException.java
+++ b/jetty-util/src/main/java/org/eclipse/jetty/util/MultiException.java
@@ -31,11 +31,28 @@ import java.util.List;
 public class MultiException extends Exception
 {
     private List<Throwable> nested;
+    boolean superConstructorCalled = false;
 
     /* ------------------------------------------------------------ */
     public MultiException()
     {
         super("Multiple exceptions");
+        this.superConstructorCalled = true;
+    }
+    
+    /**
+     * We only want to fill in the stack trace if we have an exception to throw.
+     * 
+     * This prevents the stack trace from being written when the constructor is
+     * called, yet will let as fill it in later on e.g. when we actually want
+     * to throw the multi exception.
+     */
+    @Override
+    public Throwable fillInStackTrace() {
+        if(superConstructorCalled) {
+            return super.fillInStackTrace();
+        }
+        return this;
     }
 
     /* ------------------------------------------------------------ */
@@ -105,6 +122,7 @@ public class MultiException extends Exception
               if (th instanceof Exception)
                   throw (Exception)th;
           default:
+              this.fillInStackTrace();
               throw this;
         }
     }
@@ -137,6 +155,7 @@ public class MultiException extends Exception
               else
                   throw new RuntimeException(th);
           default:
+              this.fillInStackTrace();
               throw new RuntimeException(this);
         }
     }
@@ -155,7 +174,10 @@ public class MultiException extends Exception
             return;
         
         if (nested.size()>0)
+        {
+            this.fillInStackTrace();
             throw this;
+        }
     }
 
     /* ------------------------------------------------------------ */
diff --git a/jetty-util/src/test/java/org/eclipse/jetty/util/MultiExceptionTest.java b/jetty-util/src/test/java/org/eclipse/jetty/util/MultiExceptionTest.java
index 091d4373fc..44d7e1af5b 100644
--- a/jetty-util/src/test/java/org/eclipse/jetty/util/MultiExceptionTest.java
+++ b/jetty-util/src/test/java/org/eclipse/jetty/util/MultiExceptionTest.java
@@ -21,10 +21,10 @@ package org.eclipse.jetty.util;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
 
-import java.io.IOException;
-
 import org.junit.Test;
 
+import java.io.IOException;
+
 
 public class MultiExceptionTest
 {
@@ -37,6 +37,8 @@ public class MultiExceptionTest
         me.ifExceptionThrow();
         me.ifExceptionThrowMulti();
         me.ifExceptionThrowRuntime();
+        
+        assertEquals("Stack trace should not be filled out", 0, me.getStackTrace().length);
     }
 
     @Test
@@ -91,19 +93,26 @@ public class MultiExceptionTest
         {
             assertTrue(run==e);
         }
+        
+        assertEquals("Stack trace should not be filled out", 0, me.getStackTrace().length);
     }
 
-    @Test
-    public void testTwo() throws Exception
-    {
+    private MultiException multiExceptionWithTwo() {
         MultiException me = new MultiException();
         IOException io = new IOException("one");
         RuntimeException run = new RuntimeException("one");
         me.add(io);
         me.add(run);
-
         assertEquals(2,me.size());
-
+        
+        assertEquals("Stack trace should not be filled out", 0, me.getStackTrace().length);
+        return me;
+    }
+    
+    @Test
+    public void testTwo() throws Exception
+    {
+        MultiException me = multiExceptionWithTwo();
         try
         {
             me.ifExceptionThrow();
@@ -112,8 +121,10 @@ public class MultiExceptionTest
         catch(MultiException e)
         {
             assertTrue(e==me);
+            assertTrue(e.getStackTrace().length > 0);
         }
 
+        me = multiExceptionWithTwo();
         try
         {
             me.ifExceptionThrowMulti();
@@ -122,8 +133,10 @@ public class MultiExceptionTest
         catch(MultiException e)
         {
             assertTrue(e==me);
+            assertTrue(e.getStackTrace().length > 0);
         }
 
+        me = multiExceptionWithTwo();
         try
         {
             me.ifExceptionThrowRuntime();
@@ -132,8 +145,10 @@ public class MultiExceptionTest
         catch(RuntimeException e)
         {
             assertTrue(e.getCause()==me);
+            assertTrue(e.getStackTrace().length > 0);
         }
 
+        RuntimeException run = new RuntimeException("one");
         me = new MultiException();
         me.add(run);
         me.add(run);
@@ -146,6 +161,7 @@ public class MultiExceptionTest
         catch(RuntimeException e)
         {
             assertTrue(e.getCause()==me);
+            assertTrue(e.getStackTrace().length > 0);
         }
     } 
```

I does require that no one is doing `throw new MultiException();` because in that case the stack will not be filled in.

**Comment by olamy** [comment]
@LukeButtersFunnelback it might be better create a pull request. Due to Eclipse IP we are not allowed to accept contributions in such format. And btw it will be easier to read/comment etc.. :-)
Thanks!

**Comment by LukeButtersFunnelback** [comment]
I made a pull request https://github.com/eclipse/jetty.project/pull/2582

**Comment by joakime** [comment]
This has been fixed since jetty-9.4.11.v20180605
